baseclasses/utils/solverHistory.py

```diff
@@ -0,0 +1,354 @@
+            headerString += self.variables[var]["headerFormat"].format(var)
+            headerString += "|"
+        headerWidth = len(headerString)
+
+        headerBar = "+" + "-" * (headerWidth - 2) + "+"
+        print(headerBar)
+        print(headerString)
+        print(headerBar)
+
+    def printData(self, iters: Optional[Union[int, Iterable[int]]] = None) -> None:
+        """Print a selection of lines from the history
+
+        Parameters
+        ----------
+        iters : int or Iterable of ints, optional
+            Iteration numbers to print, by default only the last iteration will be printed
+        """
+        if iters is None:
+            iters = [-1]
+        elif isinstance(iters, int):
+            iters = [iters]
+
+        if max(iters) >= self.iter:
+            raise ValueError(
+                f"Requested iteration {max(iters)} is not in the history, only {self.iter} iterations in history"
+            )
+        if min(iters) < -self.iter:
+            raise ValueError(
+                f"Requested iteration {min(iters)} is not in the history, only {self.iter} iterations in history"
+            )
+
+        for i in iters:
+            lineString = "|"
+            for var in self.variablesToPrint:
+                if self.data[var][i] is None:
+                    lineString += self.variables[var]["headerFormat"].format("-")
+                else:
+                    valueString = self.variables[var]["format"].format(self.data[var][i])
+                    lineString += self.variables[var]["headerFormat"].format(valueString)
+                lineString += "|"
+            print(lineString)
+
+    def save(self, fileName: str) -> None:
+        """Write the solution history to a pickle file
+
+        Only the data dictionary is saved
+
+        Parameters
+        ----------
+        fileName : str
+            File path to save the solution history to, file extension not required, will be ignored if supplied
+        """
+        base = os.path.splitext(fileName)[0]
+        fileName = base + ".pkl"
+
+        dataToSave = {"data": self.getData(), "metadata": self.getMetadata()}
+        with open(fileName, "wb") as file:
+            pickle.dump(dataToSave, file, protocol=pickle.HIGHEST_PROTOCOL)
+
+    def getData(self) -> Dict[str, List]:
+        """Get the recorded data
+
+        Returns
+        -------
+        dict
+            Dictionary of recorded data
+        """
+        return copy.deepcopy(self.data)
+
+    def getMetadata(self) -> Dict[str, Any]:
+        """Get the recorded metadata
+
+        Returns
+        -------
+        dict
+            Dictionary of recorded metadata
+        """
+        return copy.deepcopy(self.metadata)
+
+    def getVariables(self) -> List[str]:
+        """Get the recorded variables
+
+        Returns
+        -------
+        dict
+            Dictionary of recorded variables
+        """
+        return list(self.variables.keys())
+
+    @property
```

Comment:
Why use a `property` here?

Reply:
See my response to one of Marco's comments above. I use a property because I don't know at which point to recompute this list otherwise. You never know when the user has stopped adding variables.

Reply:
Right, but I guess what I had in mind is why use the `property` decorator here as opposed to just a regular function. They are typically used instead of get/setters for member variables. This one is getting variable names to be printed. The `getVariables` and `getMetadata` etc. seem to functions that could be used with property.

Reply:
What do you think of the new implementation?

Reply:
I guess I used a property for this as it's kinda used as a private attribute of the class, whereas `getVariables`, `getData` and `getMetadata` are all public methods

Reply:
Not an issue. Its somewhat of a style and in some cases makes sense and is convenient. Get and set functions are fine.